File: 06_Linked_List/01_Reverse_Linked_List/0206-reverse-linked-list.cpp
/*
Problem: LeetCode 206 - Reverse Linked List

Description:
Reverse a singly linked list.

Intuition:
To reverse a linked list, we need to change the direction of each node's "next" pointer. We can do this iteratively by keeping track of three nodes: current, previous, and next. We start with current pointing to the head of the original list, previous as nullptr (since there is no node before the head), and next as the next node of the current node. During each iteration, we update the "next" pointer of the current node to point to the previous node. Then, we move the previous node to be the current node and the current node to be the next node. We repeat this process until the end of the original list is reached.

Approach:
1. Create three pointers: current, previous, and next.
2. Initialize current to the head of the original list and previous as nullptr.
3. While the current node is not nullptr:
   a. Update the "next" pointer of the current node to point to the previous node.
   b. Move previous to be the current node and current to be the next node.
4. At the end of the loop, the previous node will be the new head of the reversed list.

Time Complexity:
The time complexity is O(N), where N is the number of nodes in the linked list. We visit each node once during the reversal process.

Space Complexity:
The space complexity is O(1) since we are using a constant amount of extra space for the three pointers.
*/

class Solution {
  public:
    ListNode *reverseList(ListNode *head) {
        ListNode *current = head;
        ListNode *previous = nullptr;
        ListNode *next;

        while (current) {
            next = current->next;
            current->next = previous;
            previous = current;
            current = next;
        }

        return previous;
    }
};

File: 06_Linked_List/01_Reverse_Linked_List/0206-reverse-linked-list.py
"""
Problem: LeetCode 206 - Reverse Linked List

Key Idea:
To reverse a singly linked list, we need to reverse the direction of the pointers while traversing the list. We maintain three pointers: 'prev' (to keep track of the previous node), 'current' (to keep track of the current node), and 'next_node' (to keep track of the next node in the original list). In each iteration, we update the 'current.next' pointer to point to the 'prev' node, and then move 'prev' and 'current' pointers one step forward. We repeat this process until we reach the end of the original list, and the 'prev' pointer will be pointing to the new head of the reversed list.

Time Complexity:
The time complexity of this solution is O(n), where n is the number of nodes in the linked list. We traverse each node once to reverse the list.

Space Complexity:
The space complexity is O(1), as no extra space is used other than a few variables to keep track of nodes and pointers.
"""

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next


class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        prev = None
        current = head

        while current:
            next_node = current.next
            current.next = prev
            prev = current
            current = next_node

        return prev
